Verifier simplification.

timothee-haudebourg opened the pull request:
Currently the `VerifiableClaims::verify` function (or more precisely `verify_with`) takes a "verifier" and an "environment". The "verifier" is in fact a public key resolver (W3C verification method resolver or JWK resolver, etc), while the environment provides any other resource required to validate the claims and signature.

I realized there is no real reason to separate the resolver from the environment. Merging them into a single `verifier` allows us to remove an input argument to many functions (including `VerifiableClaims::verify`) and remove a type parameter to some traits. This is the purpose of this PR.

Here is an overview of the changes:
  - Remove the `environment` argument in `VerifiableClaims::verify`, the `verifier` is now the "environment".
  - Remove the `VerifiableClaims::verify_with`, now unnecessary.
  - Add a `ResolverEnvironment` trait implemented by any type providing a public key resolver (similar to other `*Environment` traits).
  - Add a `Verifier` type, implementing `ResolverEnvironment` and all the commonly used `*Environment` traits. This is the default built-in verifier type that works with most verifiable claims. It replaces the old `VerificationEnvironment` type.
  - Rename `Validate` into `ValidateClaims`.
  - Remove the `verifier` argument from `ValidateProof`. Now `ValidateClaims` and `ValidateProof` are completely symmetrical.
  - Rename `JWSVerifier` into `JWKResolver`. This makes the function of this trait clearer: its a type that can resolve a key id into a JWK. Just like `VerificationMethodResolver` resolves a key id into a W3C verification method.

The only downside is that the `verify` function must take an actual verifier as parameter, and not just the public key resolver. A verifier can be built from a resolver using `Verifier::from_resolver`. It's one more step, but I also think it makes more sense while making customizing the verifier easier. For instance its possible to customize the JSON-LD context loader while constructing the verifier with one line:
```
// Create verifier with custom LD loader.
let verifier = Verifier::from_resolver(my_resolver).with_json_ld_loader(my_custom_loader);

// Verify the claims.
vc.verify(&verifier).await;
```
Before, you would need to construct your own verification environment and use `verify_with` instead of `verify`.

timothee-haudebourg commented:
Also, I renamed the `DIDResolver::with_default_options` function into `into_vm_resolver` to make it clear that it turns a DID resolver into a verification method resolver.

sbihel commented on crates/claims/core/src/verification/proof.rs:
```diff
@@ -108,7 +108,7 @@ impl From<std::convert::Infallible> for ProofValidationError {
 
 /// Proof that can be validated against `T` claims with a verifier of type `V`.
-pub trait ValidateProof<T, E, V> {
+pub trait ValidateProof<V, T> {
```
Is there some guidance from the specs on the distinction between verification and validation? IMO validation is for "static checks" (e.g. making sure it's well formatted) and verification is a superset of validation. But then you can say that a signature is "valid", so I don't know.

timothee-haudebourg replied:
As far as I understand validation is checking that the claims themselves are compatibles with the verifier. That can mean checking the claims are not expired, the issuer is supported, the verifier is part of the claims audience, etc. Well formlessness checks that should be handled by the type system as much as possible, but more complex well-formlessness properties could also be checked at runtime during validation.

But as you said, the specs also often use the term "valid" for signatures. That's why to conciliate the two notions I've introduce the two traits `ValidateClaims` for the former understanding regarding claims, and `ValidateProof` for the later regarding signatures. Then "verifying" consists in validating the claims AND the signature.

sbihel commented on crates/claims/core/src/verification/mod.rs:
```diff
@@ -98,3 +93,135 @@ pub enum Invalid {
 }
+
+/// Public key resolver environment.
+pub trait ResolverEnvironment {
```
I don't know if environment is the right term. In SIWE for example, you provide "verification parameters" which can include a different time from now

timothee-haudebourg replied:
Yes that's the same idea with the `*Environment` traits. For now we have `DateTimeEnvironment`, `ResolverEnvironment`, `ContextLoaderEnvironment` and `Eip712TypesEnvironment`, and probably more in the future. I agree "Environment" may not be the best term, but I couldn't come up with a better one yet. Maybe "Provider".

sbihel replied:
After talking about it with Jacob, we decided to use `Provider` instead of `Environment`

timothee-haudebourg commented:
I've renamed all the `*Environment` traits into `*Provider`, added a dedicated `verify` function for each secured claim type (`CompactJWS`, `DecodedJWS` and `DataIntegrity`), renamed `Verifier` into `VerificationParameters` and added more documentation about verification parameters.
